### ISSUE TITLE
support searches across multiple projects (Bugzilla #341)

### ISSUE THREAD

**Issue by vladak** [oracle/opengrok#330]
status CLOSED severity _enhancement_ in component _webapp_ for _0.8_
Reported in version _unspecified_ on platform _Other_
Assigned to: Lubos Kosco

On 2008-01-04 11:44:35 +0000, Hemang Lavana wrote:

> Hi,
> 
> There was a recent thread discussing the need for supporting search across multiple projects:
> 
> http://www.opensolaris.org/jive/thread.jspa?threadID=48141&tstart=0
> 
> It seems that the code in trunk creates separate index for each project. The primary benefit of doing this is that the spell checker will not provide suggestions for search terms outside of the defined project.
> 
> On the other hand, creating separate index per project makes it difficult to search across multiple projects. A simple approach that was suggested in the thread would be to modify search.jsp to have a loop that repeats the search for each selected project. However, this might result in increased time required to search the keyword by number of projects and it may also skew the results. For example, the search keyword may occur only once in a given project and that result may still appear in the overall result which doesn't seem correct.
> 
> This is RFE to support searches across multiple projects.
> 
> Hemang.

On 2008-02-23 12:30:29 +0000, Lubomir Rintel wrote:

> Search across all/multiple project is my most frequent use case (list of projects containing thousand projects). Still I can use the path matching for that, it's just not as well integrated into the user interface.

On 2008-08-26 01:24:03 +0000, Trond Norbye wrote:

> **\* Bug 3068 has been marked as a duplicate of this bug. ***

On 2008-08-31 03:00:05 +0000, Trond Norbye wrote:

> I suggest that we implement this RFE in two steps. The first step will be to loop through the selected projects and search all of them (this is a small quick fix), and the second part will be to merge all the project index databases into a single database and search that database similar to the way we did before..

On 2009-06-22 17:11:09 +0000, Garen wrote:

> fwiw, my initial motivation to use opengrok was to be able to help with finding security vulnerabilities.
> 
> When a new high profile open-source security vulnerability is announced, I want to be able to search with opengrok for some string that I find in the affected code and quickly verify if it is in the codebase indexed by opengrok.  I don't want to have to monotonously select each project one by one.
> 
> The other common case I want it for is for finding all instances of a particular defect -- find all the sources that re-used a portion of 'defective' code -- and make sure they all get fixed.
> 
> Need to be able to search the entire codebase for both of those use cases.

On 2009-08-21 15:16:30 +0000, Lubos Kosco wrote:

> **\* Bug 10758 has been marked as a duplicate of this bug. ***

On 2009-09-01 07:27:10 +0000, Lubos Kosco wrote:

> will fix, since new lucene update provides methods to achieve this quite conveniently, even with parallel search if more cpus are available ;)
> 
> I have some doubts on spellchecker though, but first idea is to have spellchecker suggestion per project, so it will show:
> project blah , have you meant: blah ?
> project foo , have you meant: bar ?
> project test , have you meant: testme ?
> underneath a failed search for the appropriate project, obviously results from the rest of the projects will be shown below suggestions
> 
> the other change will be that project list will move to right and a checkbox "all" will be added

On 2009-09-04 14:47:32 +0000, Lubos Kosco wrote:

> OK, changeset 830 f38eb240ab54 has the draft changes
> 
> I have to note, that the spellchecker will kick in only when nothing found in ANY project (by design from old design) .
> 
> Also I will not close this bug, since I'd like to redesign how the search page looks like to make it a bit more usable then current one ...

On 2009-10-06 09:28:58 +0000, Lubos Kosco wrote:

> created new bug for UI redesign, closing this one, most probably 0.8 will stay like it is in current trunk

On 2009-10-06 19:45:21 +0000, Lubos Kosco wrote:

> almost forgot, OSR for this is # 12974

On 2009-10-06 19:46:23 +0000, Lubos Kosco wrote:

> (In reply to comment # 9)
> 
> > almost forgot, OSR for this is # 12974
> 
> I mean the OSR for jquery used for autocomplete project input field :)
